README
NAME
    Data::NDS::Multifile - A multi-file version of Data::NDS::Multiele

SYNOPSIS
      use Data::NDS::Multifile;

DESCRIPTION
    This module is an extension to the Data::NDS::Multiele module, and an
    understanding of that module is assumed in this documentation.

    The Data::NDS::Multiele module accesses a set of data elements in a
    single data source. This module adds the ability to treat a set of data
    elements from multiple files as one set of elements.

    There are some limitations to this functionality that are not present in
    the Data::NDS::Multiele module:

    Ordered lists are not supported
        Separating ordered elements across files makes several operations
        difficult, so this is currently not supported. Only unordered lists
        are used.

    Default elements only apply to the file
        A default element in one file ONLY applies to data elements in that
        file. If the same default is required in other files, the data
        element must be added to each file.

    All files must be identical in many ways
        All will use the same Data::NDS object, so they will have identical
        structural characteristics. All files must contain hashes OR list,
        not a mixture of the two.

    All elements must be unique
        For hash files, the same element may not appear in multiple files.

    Default elements must be unique
        Since every element must be unique, the same name may not be used
        for default elements in two different files. If there is a default
        element in one file named _default, an element of the same name may
        not exist in a second file.

    List element indexes differ
        Since multiple files are read in, the index of list elements is the
        index in the combined list of elements with all of the elements from
        all files.

METHODS
    new
    file
        A new Data::NDS::Multifile object can be created and assigned to
        data sources using these methods. In order to create a new object
        and bind it to two different files, use the following calls:

           $obj = new Data::NDS::Multifile [NDS];
           $obj->file(LABEL1,FILE1);
           $obj->file(LABEL2,FILE2);

        or an equivalent one call method:

           $obj = new Data::NDS::Multifile [NDS,] LABEL1, FILE1, LABEL2, FILE2;

        A Data::NDS object is needed. If one is passed in as an argument to
        the new method, it is used. By creating an NDS, and then passing it
        in to the new method, the same NDS can be used for multiple
        Data::NDS::Multiele/Data::NDS::Multifile objects, each reading
        elements independent of each other. In this way, all structural
        information in the NDS applies to all elements use by all of the
        objects.

        If no Data::NDS object is passed in, one is created.

        Creating two Data::NDS::Multifile objects which use the same NDS can
        be done in the following way:

           $NDS  = new Data::NDS;
           $obj1 = new Data::NDS::Multifile $NDS [LABEL1a, FILE1a, ...];
           $obj2 = new Data::NDS::Multifile $NDS [LABEL2a, FILE2a, ...];

        The same effect can be achieved with:

           $obj1 = new Data::NDS::Multifile [LABEL1a, FILE1a, ...];
           $obj2 = $obj1->new([LABEL2a, FILE2a, ...]);

        If no file is passed in to the new method, you need to use the file
        method to set it and read the data.

        One note about labels: they should not be strictly numeric since
        there are times when that may conflict with numeric element indices
        in data consisting of lists. Add any non-digit character to them to
        avoid this problem.

        All file calls must be done immediately after the object is created,
        and before any other operations (including default_element
        oeprations).

    nme
           $NME = $obj->nme(LABEL);

        This returns the Data::NDS::Multiele object associated with a label.

    ele_file
           $label = $obj->ele_file($ele);

        Returns the label for the file that $ele belongs to. Returns nothing
        if the element does not exist.

    default_element
           $obj->default_element($label [,$ruleset] [,$path,$val,...]);
           $obj->default_element($ele   [,$ruleset] [,$path,$val,...]);

        This is used to declare that one of the elements that was read in
        from one of the data file is used to provide defaults for other
        elements in that same data file.

        If the data files contains lists of elements, the first form is used
        ($label must be passed in as the first argument). The first element
        from the given file is used as a default.

        If the data files contain hashes of elements, the second form is
        used ($ele is passed in as the first argument). The named element
        (in whichever file it occurs) is used as a default. It should be
        noted that default elements must be uniquely named in a Multifile
        object.

        The default_element methods must be done immediately after the file
        methods, and before any other operations.

    rename_ele
           $obj->rename_ele($ele,$newele);

        This is identical to the function in Data::NDS::Multiele. The only
        thing to note is that the renamed element will remain in the same
        file. The only way to move an element from one file to another is to
        first delete it, and then add it in the new file.

        The rename_ele method is not applicable to a file containing a list
        of elements since the order of the elements is not currently used,
        and ordered lists are not supported.

    add_ele
           $obj->add_ele([$label,] [$ele,] $nds [,$new]);

        This adds a new element to a file. The new element is checked for
        validity, and then added. Note that the element is not actually
        stored in the data file. Use the save method to do that.

        Because the NDS does not have defaults applied, the new element may
        have additional data present once defaults are applied.

        The location (and file) where the new element is added depends on
        several things.

        If the data files contain lists, there are three different ways to
        call this:

           $obj->add_ele([$label,] $nds [,$new]);

        This adds the element to the end of the given file, or the last data
        file if $label is not included.

           $obj->add_ele($ele,$nds [,$new]);

        This inserts the new element at $ele (which must refer to an
        existing element) in the same data file as $ele.

        If the data file contains a hash, there is only one way to call it:

           $obj->add_ele([$label,] $ele,$nds [,$new]);

        If $label is not given, it is inserted into the last data file.
        Otherwise, it is inserted into the file specifed by $label.

    copy_ele
           $obj->copy_ele($ele [,$label] [,$newele]);

        This will create a new element which is a copy of another element.
        The new element is created in the same way as the add_ele method
        based on the value of $newele and the type of data.

        If $label is not given, it defaults to the same file as the original
        element.

    version
    nds
    err
    errmsg
    eles
    ele
    which
    path_valid
    value
    keys
    values
    path_values
    path_in_use
    delete_ele
    update_ele
    is_default_value
    dump
    save
        These work the same as in the Data::NDS::Multiele module and are
        fully documented in that module.

    Any method not documented here, especially those beginning with an
    underscore (_), are for internal use only. Please do not use them.
    Absolutely no support is offered for them.

ERROR CODES
    Each error code produced by a method in the Data::NDS::Multifile module
    is prefixed by the characters "nmf", followed by a 3 character operation
    code which tells what type of operation failed, followed by 2 digits.

    The following error codes are used to identify problems working with
    files containing data:

       nmffil01   An even number of arguments required to
                  specify files.
       nmffil02   An attempt to reuse a file label already in use.
       nmffil03   An error occurred reading the data file.
       nmffil04   All files must contain the same type of data (list
                  or hash).
       nmffil05   A data element is duplicated in 2 files.
       nmffil06   An invalid file label was used.
       nmffil07   An attempt to read in a file after element
                  operations have been done.
       nmffil08   No file has been set.
       nmffil09   Defaults must be set immediately after the files
                  are read in.

    The following error codes identify problems accessing an element:

       nmfele01   Attempt to access an undefined element.
       nmfele02   Attempt to overwrite an existing element.
       nmfele03   When accessing a hash element, a name must be given.
       nmfele04   When adding an element, a name must be given.

BUGS AND QUESTIONS
    If you find a bug in this module, please send it directly to me (see the
    AUTHOR section below). Alternately, you can submit it on CPAN. This can
    be done at the following URL:

       http://rt.cpan.org/Public/Dist/Display.html?Name=Data-NDS-Multifile

    Please do not use other means to report bugs (such as usenet newsgroups,
    or forums for a specific OS or linux distribution) as it is impossible
    for me to keep up with all of them.

    When filing a bug report, please include the following information:

    *   The version of the module you are using. You can get this by using
        the script:

                use Data::NDS::Multifile;
                $obj = new Data::NDS::Multifile;
                print $obj->version(),"\n";

    *   The output from "perl -V"

    If you have a problem using the module that perhaps isn't a bug (can't
    figure out the syntax, etc.), you're in the right place. Go right back
    to the top of this manual and start reading. If this still doesn't
    answer your question, mail me directly.

KNOWN PROBLEMS
    None at this point.

LICENSE
    This script is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

AUTHOR
    Sullivan Beck (sbeck@cpan.org)



HISTORY
###############################################################################
# HISTORY
###############################################################################

Version 3.00  2009-03-02
   Initial release. Most of the functionality comes from the
      Data::NDS::Multisource module.

Version 3.10  2009-04-14
   Cleaned up some of the internals.

Version 3.11  2010-02-05
   Added path_in_use method.
   Minor bug fix in path_values.

Version 3.12  2010-03-04
   Fixed "use Data::NDS::Multiele" to require the correct version.
   Renamed the module to Data::Nested::Multifile and added to the
      Data::Nested distribution .  3.12 is the final release of
      Data::NDS::Multifile .

###############################################################################
# BACKWARDS INCOMPATIBILITIES
###############################################################################

Version 3.10
   The API of the 3.10 series is backward compatible with 3.00, but
   due to changes in the internals, the 3.10 Data::NDS module is
   required for the 3.10 version of other Data::NDS::* modules.



t/test.pl
#!/usr/bin/perl -w
# Copyright (c) 1996-2010 Sullivan Beck. All rights reserved.
# This program is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.

# SB_TEST.PL
###############################################################################
# HISTORY
#
# 1996-??-??  Wrote initial version for Date::Manip module
#
# 1996-2001   Numerous changes
#
# 2001-03-29  Rewrote to make it easier to drop in for any module.
#
# 2001-06-19  Modifications to make space delimited stuff work better.
#
# 2001-08-23  Added support for undef args.
#
# 2007-08-14  Better support for undef/blank args.
#
# 2008-01-02  Better handling of $runtests.
#
# 2008-01-24  Better handling of undef/blank args when arguements are
#             entered as lists instead of strings.
#
# 2008-01-25  Created a global $testnum variable to store the test number
#             in.
#
# 2008-11-05  Slightly better handling of blank/undef in returned values.
#
# 2009-09-01  Added "-l" value to $runtests.
#
# 2009-09-30  Much better support for references.

###############################################################################

use Storable qw(dclone);

# Usage: test_Func($funcref,$tests,$runtests,@extra)=@_;
#
# This takes a series of tests, runs them, compares the output of the tests
# with expected output, and reports any differences.  Each test consists of
# several parts:
#    a function passed in as a reference ($funcref)
#    a series of arguments to be passed to the function
#    the expected output from the function call
#
# Tests may be passed in in two methods: as a string, or as a reference.
#
# Using the string case, $tests is a newline delimited string.  Each test
# takes one or more lines of the string.  Tests are separated from each
# other by a blank line.
#
# Arguments and return value(s) may be written as a single line:
#    ARG1 ARG2 ... ARGn ~ VAL1 VAL2 ... VALm
# or as multiple lines:
#    ARG1
#    ARG2
#    ...
#    ARGn
#    ~
#    VAL1
#    VAL2
#    ...
#    VALm
#
# If any of the arguments OR values have spaces in them, only the multiline
# form may be used.
#
# If there is exactly one return value, the separating tilde is
# optional:
#    ARG1 ARG2 ... ARGn VAL1
# or:
#    ARG1
#    ARG2
#    ...
#    ARGn
#    VAL
#
# It is valid to have a function with no arguments or with no return
# value (or both).  The "~" must be used:
#
#    ARG1 ARG2 ... ARGn ~
#
#    ~ VAL1 VAL2 ... VALm
#
#    ~
#
# Leading and trailing space is ignored in the multi-line format.
#
# If desired, any of the ARGs or VALs may be the word "_undef_" which
# will be strictly interpreted as the perl undef value. The word "_blank_"
# may also be used to designate a defined but empty string.
#
# They may also be (in the multiline format) of the form:
#
#   \ STRING           : a string reference
#
#   [] LIST            : a list reference (where LIST is a
#                        comma separated list)
#
#   [SEP] LIST         : a list reference (where SEP is a
#                        single character separator)
#
#   {} HASH            : a hash reference (where HASH is
#                        a comma separated list)
#
#   {SEP} HASH         : a hash reference (where SEP is a
#                        single character separator)
#
# Alternately, the tests can be passed in as a list reference:
#    $tests = [
#               [
#                 [ @ARGS1 ],
#                 [ @VALS1 ]
#               ],
#               [
#                 [ @ARGS2 ],
#                 [ @VALS2 ]
#               ], ...
#             ]
#
# @extra are extra arguments which are added to the function call.
#
# There are several ways to run the tests, depending on the value of
# $runtests.
#
# If $runtests is 0, the tests are run in a non-interactive way suitable
# for running as part of a "make test".
#
# If $runtests is a positive number, it runs runs all tests starting at
# that value in a way suitable for running interactively.
#
# If $runtests is a negative number, it runs all tests starting at that
# value, but providing feedback at each test.
#
# If $runtests is a string "=N" (where N is a number), it runs only
# that test.
#
# If $runtests is the string "-l", it lists the tests and the expected
# output without running any.

sub test_Func {
   my($funcref,$tests,$runtests,@extra)=@_;
   my(@tests);

   $runtests     = 0  if (! $runtests);
   my($starttest,$feedback,$endtest,$runtest);
   if      ($runtests eq "0"  or  $runtests eq "-0") {
      $starttest = 1;
      $feedback  = 1;
      $endtest   = 0;
      $runtest   = 1;
   } elsif ($runtests =~ /^\d+$/){
      $starttest = $runtests;
      $feedback  = 0;
      $endtest   = 0;
      $runtest   = 1;
   } elsif ($runtests =~ /^-(\d+)$/) {
      $starttest = $1;
      $feedback  = 1;
      $endtest   = 0;
      $runtest   = 1;
   } elsif ($runtests =~ /^=(\d+)$/) {
      $starttest = $1;
      $feedback  = 1;
      $endtest   = $1;
      $runtest   = 1;
   } elsif ($runtests eq "-l") {
      $starttest = 1;
      $feedback  = 1;
      $endtest   = 0;
      $runtest   = 0;
   } else {
      die "ERROR: unknown argument(s): $runtests";
   }

   if (ref($tests) eq "ARRAY") {
      @tests = @$tests;

   } else {
      # Separate tests.

      my($comment)="#";
      my(@lines)=split(/\n/,$tests);
      my(@test);
      while (@lines) {
         my $line = shift(@lines);
         $line =~ s/^\s*//;
         $line =~ s/\s*$//;
         next  if ($line =~ /^$comment/);

         if ($line ne "") {
            push(@test,$line);
            next;
         }

         if (@test) {
            push(@tests,[ @test ]);
            @test=();
         }
      }
      if (@test) {
         push(@tests,[ @test ]);
      }

      # Get arg/val lists for each test.

      foreach my $test (@tests) {
         my(@tmp)=@$test;
         my(@arg,@val);

         # single line test
         @tmp = split(/\s+/,$tmp[0])  if ($#tmp == 0);

         my($sep)=-1;
         my($i);
         for ($i=0; $i<=$#tmp; $i++) {
            if ($tmp[$i] eq "~") {
               $sep=$i;
               last;
            }
         }

         if ($sep<0) {
            @val=pop(@tmp);
            @arg=@tmp;
         } else {
            @arg=@tmp[0..($sep-1)];
            @val=@tmp[($sep+1)..$#tmp];
         }
         $test = [ [@arg],[@val] ];
      }
   }

   my($ntest)=$#tests + 1;
   print "1..$ntest\n"  if ($feedback  &&  $runtest);

   my(@t);
   if ($endtest) {
      @t = ($starttest..$endtest);
   } else {
      @t = ($starttest..$ntest);
   }

   foreach my $t (@t) {
      $::testnum  = $t;

      my $arg     = dclone($tests[$t-1][0]);
      my @arg     = @$arg;
      print_to_vals(\@arg);

      my $argprt  = dclone(\@arg);
      my @argprt  = @$argprt;
      vals_to_print(\@argprt);

      my $exp     = dclone($tests[$t-1][1]);
      my @exp     = @$exp;
      print_to_vals(\@exp);
      vals_to_print(\@exp);

      # Run the test

      my ($ans,@ans);
      if ($runtest) {
         @ans = &$funcref(@arg,@extra);
      }
      vals_to_print(\@ans);

      # Compare the results

      foreach my $arg (@arg) {
         $arg = "_undef_"  if (! defined $arg);
         $arg = "_blank_"  if ($arg eq "");
      }
      $arg = join("\n           ",@argprt,@extra);
      $ans = join("\n           ",@ans);
      $exp = join("\n           ",@exp);

      if (! $runtest) {
         print "########################\n";
         print "Test     = $t\n";
         print "Args     = $arg\n";
         print "Expected = $exp\n";
      } elsif ($ans ne $exp) {
         print "not ok $t\n";
         warn "########################\n";
         warn "Args     = $arg\n";
         warn "Expected = $exp\n";
         warn "Got      = $ans\n";
         warn "########################\n";
      } else {
         print "ok $t\n"  if ($feedback);
      }
   }
}

# The following is similar but it takes input from an input file and
# sends output to an output file.
#
# $files is a reference to a list of tests.  If one of the tests is named
# "foobar", the input is from "foobar.in", output is to "foobar.out", and
# the expected output is in "foobar.exp".
#
# The function stored in $funcref is called as:
#    &$funcref($in,$out,@extra)
# where $in is the name of the input file, $out is the name of the output
# file, and @extra are any additional arguments that are required.
#
# The function should return 0 on success, or an error message.

sub test_File {
   my($funcref,$files,$runtests,@extra)=@_;
   my(@files)=@$files;

   $runtests=0  if (! $runtests);

   my($ntest)=$#files + 1;
   print "1..$ntest\n"  if (! $runtests);

   my(@t);
   if ($runtests > 0) {
      @t = ($runtests..$ntest);
   } elsif ($runtests < 0) {
      @t = (-$runtests);
   } else {
      @t = (1..$ntest);
   }

   foreach my $t (@t) {
      $::testnum = $t;
      my $test = $files[$t-1];
      my $expf = "$test.exp";
      my $outf = "$test.out";

      if (! -f $test  ||  ! -f $expf) {
         print "not ok $t\n";
         warn  "Test: $test: missing input/outpuf information\n";
         next;
      }

      my $err  = &$funcref($test,$outf,@extra);
      if ($err) {
         print "not ok $t\n";
         warn  "Test: $test: $err\n";
         next;
      }

      local *FH;
      open(FH,$expf)  ||  do {
         print "not ok $t\n";
         warn  "Test: $test: $!\n";
         next;
      };
      my @exp = <FH>;
      close(FH);
      my $exp = join("",@exp);
      open(FH,$outf)  ||  do {
         print "not ok $t\n";
         warn  "Test: $test: $!\n";
         next;
      };
      my @out = <FH>;
      close(FH);
      my $out = join("",@out);

      if ($out ne $exp) {
         print "not ok $t\n";
         warn  "Test: $test: output differs from expected value\n";
         next;
      }

      print "ok $t\n"  if (! $runtests);
   }
}

# Converts a printable version of arguments to actual arguments
sub print_to_vals {
   my($listref) = @_;

   foreach my $arg (@$listref) {
      next  if (! defined($arg));
      if ($arg eq "_undef_") {
         $arg = undef;

      } elsif ($arg eq "_blank_") {
         $arg = "";

      } elsif ($arg =~ /^\\\s*(.*)/) {
         $str = $1;
         $arg = \$str;

      } elsif ($arg =~ /^\[(.?)\]\s*(.*)/) {
         my($sep,$str) = ($1,$2);
         $sep = ","  if (! $sep);
         my @list = split(/\Q$sep\E/,$str);
         foreach my $e (@list) {
            $e = ""     if ($e eq "_blank_");
            $e = undef  if ($e eq "_undef_");
         }
         $arg = \@list;

      } elsif ($arg =~ /^\{(.?)\}\s*(.*)/) {
         my($sep,$str) = ($1,$2);
         $sep = ","  if (! $sep);
         my %hash = split(/\Q$sep\E/,$str);
         foreach my $key (keys %hash) {
            my $val = $hash{$key};
            $hash{$key} = undef  if ($val eq "_undef_");
            $hash{$key} = ""     if ($val eq "_blank_");
         }
         $arg = \%hash;
      }
   }
}

# Converts arguments to a printable version.
sub vals_to_print {
   my($listref) = @_;

   foreach my $arg (@$listref) {
      if (! defined $arg) {
         $arg = "_undef_";

      } elsif (! ref($arg)) {
         $arg = "_blank_"  if ($arg eq "");

      } else {
         my $ref = ref($arg);
         if      ($ref eq "SCALAR") {
            $arg = "\\ $$arg";

         } elsif ($ref eq "ARRAY") {
            my @list = @$arg;
            foreach my $e (@list) {
               $e = "_undef_", next   if (! defined($e));
               $e = "_blank_"         if ($e eq "");
            }
            $arg = join(" ","[",join(", ",@list),"]");

         } elsif ($ref eq "HASH") {
            %hash = %$arg;
            foreach my $key (keys %hash) {
               my $val = $hash{$key};
               $hash{$key} = "_undef_", next  if (! defined($val));
               $hash{$key} = "_blank_"        if ($val eq "_blank_");
            }
            $arg = join(" ","{",
                        join(", ",map { "$_ => $hash{$_}" }
                             (sort keys %hash)), "}");
            $arg =~ s/  +/ /g;
         }
      }
   }
}

1;
# Local Variables:
# mode: cperl
# indent-tabs-mode: nil
# cperl-indent-level: 3
# cperl-continued-statement-offset: 2
# cperl-continued-brace-offset: 0
# cperl-brace-offset: 0
# cperl-brace-imaginary-offset: 0
# cperl-label-offset: -2
# End:



Build.PL
use Module::Build;
use strict;
use warnings;

my %requires = (
                 'perl'                     => '5.001',
                 'Data::NDS'                => '0',
                 'Data::NDS::Multiele'      => '0',
                 'Storable'                 => '0',
                 'YAML::Syck'               => '0',
               );
my %build_mods = (
                 'Test::More'               => '0',
                 'Test::Pod'                => '0',
                 'Test::Pod::Coverage'      => '0',
               );


my $build = Module::Build->new(
        license            => 'perl',
        dist_version       => '3.12',
        dist_author        => 'Sullivan Beck <sbeck@cpan.org>',
        module_name        => 'Data::NDS::Multifile',
        dist_abstract      => 'A multi-file version of Data::NDS::Multiele',
        requires           => \%requires,
        build_requires     => {},
        build_recommends   => \%build_mods,
        sign               => 1,
);

$build->create_build_script;


Makefile.PL

use ExtUtils::MakeMaker;
use 5.001;
use strict;
use warnings;
my %requires = (
                 'Data::NDS'                => '0',
                 'Data::NDS::Multiele'      => '0',
                 'Storable'                 => '0',
                 'YAML::Syck'               => '0',
                 'Test::More'               => '0',
                 'Test::Pod'                => '0',
                 'Test::Pod::Coverage'      => '0',
               );


WriteMakefile(
        NAME             => 'Data::NDS::Multifile',
        VERSION          => '3.12',
        ($] >= 5.001
        ? (ABSTRACT=>'A multi-file version of Data::NDS::Multiele',
           AUTHOR  =>'Sullivan Beck (sbeck@cpan.org)')
        : ()),
        'dist' => {COMPRESS=>'gzip',SUFFIX=>'gz'},
        PL_FILES  => {},
        PREREQ_PM => \%requires,
);
